Cleanup cmake files

rhaschke opened the pull request:
This addresses #10 and several other minor issues in the cmake files. See individual commits and their commit messages for details.
One open question is whether Eigen and numpy should be really package dependencies [here](https://github.com/ipab-slmc/pybind11_catkin/pull/11/commits/ff0de68290a1c22d1d154ea056ef6db0b536f392#diff-34ae68f4adad56c25c5bc05dcb64794eR13-R14)?
These are only required if you actually want to use those features in your downstream project.
Hence, I would argue, the downstream project is responsible for including them.

rhaschke committed:
remove Eigen includes + Python libs from catkin_package()

These are not find_packaged and thus the corresponding variables are not defined.

rhaschke committed:
declare dependencies as build-export dependencies

They are not needed to build this package, but by downstream packages.

rhaschke committed:
Avoid extra file copy

but directly create module .so in target folder.

rhaschke committed:
Remove duplicate install file

rhaschke committed:
Update pybind11 version to 2.4.3

rhaschke committed:
Use single catkin_package()

The distinction between install and devel space was brittle and is not required:
devel-space includes are made relative to intall space by catkin

rhaschke committed:
Cleaning up variable definitions

Setting an already cached variable again is a no-op.
Thus, most of the variable settings were non-sense.

pybind11 should use the same python version as ROS.
Thus considering PYBIND_PYTHON_EXECUTABLE doesn't make sense to me.

To directly use pybind11_add_module(), it is important to correctly set PYBIND11_INCLUDE_DIR.

pybind11_catkin_INCLUDE_DIRS should include PYTHON_INCLUDE_DIRS,
maybe EIGEN3_INCLUDE_DIR?

rhaschke commented:
Looks like Travis is broken on Melodic. Looking into this now...

rhaschke committed:
ensure pybind11 uses same python version as catkin

rhaschke commented:
The Travis issue was unrelated to my changes: `industrial_ci` installs `python3` via `lsb-release` in its base docker image. This caused `pybind11` to use `python3` instead of `python2`. I added another commit to ensure that the same python version is used as by `catkin`.

wxmerkt commented on CMakeLists.txt:
```diff
@@ -18,28 +19,19 @@ ExternalProject_Add_Step(pybind11_src CopyToDevel
-    CFG_EXTRAS pybind11_catkin.cmake
-  )
-endif()
+file(MAKE_DIRECTORY ${CATKIN_DEVEL_PREFIX}/include/${PROJECT_NAME})
```
I think this may fail on the ROS buildfarm (no root access during the first build) which is why I added the weird if-logic in the first place.

rhaschke replied:
I don't see why. A normal user should have write access to CATKIN_DEVEL_PREFIX. Actually, pybind11 is installed into a subfolder of it.

wxmerkt commented:
Thank you for the clean-up!

> Thus considering PYBIND_PYTHON_EXECUTABLE doesn't make sense to me.

We have users who under ROS1 use Python3 (and manually pip3 install the rospy, rospack dependencies). Is there another way to have ROS1 with Python3 locally that I missed?

rhaschke commented:
> We have users who under ROS1 use Python3 (and manually pip3 install the rospy, rospack dependencies). Is there another way to have ROS1 with Python3 locally that I missed?

The way to select the python version for ROS is to set the environment variable `ROS_PYTHON_VERSION`, isn't it?
https://github.com/ros/catkin/blob/15df8d27136976aa2240eae9aa232728887aebb8/cmake/python.cmake#L3-L4

wxmerkt commented:
> The way to select the python version for ROS is to set the environment variable `ROS_PYTHON_VERSION`, isn't it?
> https://github.com/ros/catkin/blob/15df8d27136976aa2240eae9aa232728887aebb8/cmake/python.cmake#L3-L4

Ah, yes, that is what I had missed and was looking for.

wxmerkt commented:
Which project are you testing and building with these changes? I tried `exotica_python` locally and it doesn't work using these updates. I am looking for an example to amend accordingly

rhaschke commented:
> Which project are you testing and building with these changes? I tried `exotica_python` locally and it doesn't work using these updates. I am looking for an example to amend accordingly

I'm in the process of migrating https://github.com/ubi-agni/moveit_task_constructor/tree/wip-python-api from `boost::python` to `pybind11`. So far, only moveit_task_constructor_core is building.

rhaschke commented:
I just built exotica with the new pybind11_catkin in a clean workspace w/o issues.
What kind of problems do you observe?

wxmerkt commented:
The include directories weren't set up correctly. The minimal change to this PR that I required was:
![image](https://user-images.githubusercontent.com/1664508/71367222-5508fe00-259c-11ea-8bae-69eedb4907b4.png)

Then this PR works well for me :-)

wxmerkt commented on cmake/pybind11_catkin.cmake.in:
```diff
@@ -1,30 +1,17 @@
-
-set(PYTHON_EXECUTABLE ${TEMP_PYTHON_EXECUTABLE} CACHE INTERNAL "")
+# set variables used by pybind11_add_module()
+set(PYBIND11_INCLUDE_DIR ${pybind11_catkin_INCLUDE_DIRS})
```
For this PR to work for me, I had to change this line to 

```cmake
set(PYBIND11_INCLUDE_DIR ${pybind11_catkin_INCLUDE_DIRS}/pybind11_catkin)
```

rhaschke replied:
This should not be necessary as `pybind11_catkin_INCLUDE_DIRS` should already include `pybind11_catkin`. I will validate this with an install-space setup again...

wxmerkt replied:
My setup is a separate install space, and then a devel space that extends the install space. In this case, the `pybind11_catkin_INCLUDE_DIRS` did not include `pybind11_catkin`.

rhaschke replied:
Indeed, in an install-space setup this fails. I will have a look again. Sorry for the noise.

rhaschke replied:
Should be fixed now. @wxmerkt, could you try again please?

wxmerkt replied:
Thank you for the quick fix. This works now well for me. If you are happy, I would merge this PR and make a new bloom release to melodic and kinetic

rhaschke replied:
That would be great. Thanks.

rhaschke committed:
Handle differences between devel-space and install-space

catkin_package() doesn't export the install-space's include path correctly.
It's only named include, missing the subfolder pybind11_catkin.
